options.c

```diff
@@ -0,0 +1,454 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <getopt.h>
+
+#define KEY_VALUE_BUF_MAX 64
+#define READ_BUF_MAX 127
+
+static void
+set_api_level(char *value)
+{
+    if (!value)
+        return;
+
+    char *ptr;
+    int ret = strtol(value, &ptr, 10);
+
+    /* There is no valid number at all */
+    if (value == ptr) {
+        fprintf(stderr, "Invalid API level %s\n", value);
+        return;
+    }
+
+    /* There is a number, but also letters, this is invalid */
+    if (ptr[0] != '\0' && ptr[0] != '.') {
+        fprintf(stderr, "Invalid API level %s\n", value);
+        return;
+    }
+
+    /* This API level doesn't exist, fallback to v4 */
+    if (ret < 4)
+        ret = 4;
+
+    globalconf.api_level = ret;
+}
+
+static void
+push_arg(string_array_t *args, char *value, size_t *len)
+{
+    value[*len] = '\0';
+    string_array_append(args, a_strdup(value));
+    (*len) = 0;
+}
+
+/*
+ * Support both shebang and modeline modes.
+ */
+bool
+options_init_config(char *execpath, char *configpath, int *init_flags, string_array_t *paths)
+{
+    /* The different state the parser can have. */
+    enum {
+        MODELINE_STATE_INIT       , /* Start of line        */
+        MODELINE_STATE_NEWLINE    , /* Start of line        */
+        MODELINE_STATE_COMMENT    , /* until --             */
+        MODELINE_STATE_MODELINE   , /* until awesome_mode:  */
+        MODELINE_STATE_SHEBANG    , /* until !              */
+        MODELINE_STATE_KEY_DELIM  , /* until the key begins */
+        MODELINE_STATE_KEY        , /* until '='            */
+        MODELINE_STATE_VALUE_DELIM, /* after ':'            */
+        MODELINE_STATE_VALUE      , /* until ',' or '\n'    */
+        MODELINE_STATE_COMPLETE   , /* Parsing is done      */
+        MODELINE_STATE_INVALID    , /* note a modeline      */
+        MODELINE_STATE_ERROR        /* broken modeline      */
+    } state = MODELINE_STATE_INIT;
+
+    /* The parsing mode */
+    enum {
+        MODELINE_MODE_NONE   , /* No modeline */
+        MODELINE_MODE_LINE   , /* modeline    */
+        MODELINE_MODE_SHEBANG, /* #! shebang  */
+    } mode = MODELINE_MODE_NONE;
+
+    static const unsigned char name[] = "awesome_mode:";
+    static char key_buf [KEY_VALUE_BUF_MAX+1] = {'\0'};
+    static char file_buf[READ_BUF_MAX+1     ] = {'\0'};
+    size_t pos = 0;
+
+    string_array_t argv;
+    string_array_init(&argv);
+    string_array_append(&argv, a_strdup(execpath));
+
+    FILE *fp = fopen(configpath, "r");
+
+    /* Share the error codepath with parsing errors */
+    if (!fp)
+        return false;
+
+    /* Try to read the first line */
+    if (!fgets(file_buf, READ_BUF_MAX, fp)) {
+        fclose(fp);
+        return false;
+    }
+
+    unsigned char c;
+
+    /* Simple state machine to translate both modeline and shebang into argv */
+    for (int i = 0; (c = file_buf[i++]) != '\0';) {
+        /* Be very permissive, skip the unknown, UTF is not allowed */
```

Comment:
mb print some warning about unknown options/characters?
could be helpful in case of typo or so

Reply:
For unknown options, the reason I don't print them is a bit facepalm-ish. tl;rd, it is to let `#! /usr/bin/env awesome`  "just work" and not having to bother understanding that those 2 arguments should be ignored. Given that the executable can have any name and symlinks, correctly mapping it back to `argv[0]` is more code than this entire PR and I am very lazy.